src/SDK/SdkAutoloader.php

```diff
@@ -30,80 +36,107 @@ public static function autoload(): bool
         if (!self::isEnabled() || self::isExcludedUrl()) {
             return false;
         }
-        Globals::registerInitializer(function (Configurator $configurator) {
-            $propagator = (new PropagatorFactory())->create();
-            if (Sdk::isDisabled()) {
-                //@see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md#general-sdk-configuration
-                return $configurator->withPropagator($propagator);
-            }
-            $emitMetrics = Configuration::getBoolean(Variables::OTEL_PHP_INTERNAL_METRICS_ENABLED);
-
-            $resource = ResourceInfoFactory::defaultResource();
-            $exporter = (new ExporterFactory())->create();
-            $meterProvider = (new MeterProviderFactory())->create($resource);
-            $spanProcessor = (new SpanProcessorFactory())->create($exporter, $emitMetrics ? $meterProvider : null);
-            $tracerProvider = (new TracerProviderBuilder())
-                ->addSpanProcessor($spanProcessor)
-                ->setResource($resource)
-                ->setSampler((new SamplerFactory())->create())
-                ->build();
-
-            $loggerProvider = (new LoggerProviderFactory())->create($emitMetrics ? $meterProvider : null, $resource);
-            $eventLoggerProvider = (new EventLoggerProviderFactory())->create($loggerProvider);
-
-            ShutdownHandler::register($tracerProvider->shutdown(...));
-            ShutdownHandler::register($meterProvider->shutdown(...));
-            ShutdownHandler::register($loggerProvider->shutdown(...));
-
-            return $configurator
-                ->withTracerProvider($tracerProvider)
-                ->withMeterProvider($meterProvider)
-                ->withLoggerProvider($loggerProvider)
-                ->withEventLoggerProvider($eventLoggerProvider)
-                ->withPropagator($propagator)
-            ;
-        });
+        if (Configuration::has(Variables::OTEL_EXPERIMENTAL_CONFIG_FILE)) {
+            $sdk = self::createAndRegisterFileBasedSdk();
```

Comment:
I originally did this, but I'm coming to the conclusion that:
- if we use SPI for everything (SDK, 3rd-party + contrib implementations of APIs, and auto-instrumentations), then they are all registered in `vendor/composer/GeneratedServiceProviderData.php` as part of composer install/update
- if all services are known about in advance, we no longer have a race condition because of the undefined order of composer's `autoload.files`, where an auto-instrumentation may be registered before the SDK or some other component
- so, we can be confident that when we register instrumentations, we have already completely set up TracerProvider & friends
- so, we don't need to lazy-load them, and can pass them straight in (like in https://github.com/Nevay/opentelemetry-php/commit/2ecf4b22c0c8d76f07d6238cd75c27167d396b15#diff-c524b1bb7970cf80c06c8356ca6d3d05fb905a2b23a46aa5733489822952b556R38 - which I've retained except for creating a new class in API because of dependencies)

TL;DR: using SPI means that Globals can be immediately registered.

WDYT?

Reply:
I think my postscript above (coexisting with anything that might be using globals initializers) is what you meant here. So, I've added an initializer and a exercise adding an initializer after autoload has created an SDK, then accessing Globals.

Reply:
> TL;DR: using SPI means that Globals can be immediately registered.

Yes, if we remove `Registry::registerXxxFactory()` etc. we can replace `Globals::registerInitializer()` with a setter.

Reply:
Added a todo to `Registry` for a future enhancement, once we fully commit to SPI and get contrib converted...